File: bcs-ui/backend/templatesets/legacy_apps/configuration/validator.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
import json

from django.utils.translation import ugettext_lazy as _
from jsonschema import SchemaError
from jsonschema import ValidationError as JsonValidationError
from jsonschema import validate as json_validate
from rest_framework.exceptions import ValidationError

from .constants import KEY_PATTERN, NUM_VAR_ERROR_MSG, REAL_NUM_VAR_PATTERN
from .models import VersionedEntity, get_model_class_by_resource_name


def get_name_from_config(config):
    return config.get('metadata', {}).get('name') or ''


def is_name_duplicate(resource_name, resource_id, name, version_id):
    """同一类资源的名称不能重复"""
    # 判断新名称与老名称是否一致，如果一致，则不会重复
    model_class = get_model_class_by_resource_name(resource_name)
    try:
        resource = model_class.objects.get(id=resource_id)
        if name == resource.name:
            return False
    except model_class.DoesNotExist:
        pass

    # 只校验当前版本内是否重复
    try:
        version_entity = VersionedEntity.objects.get(id=version_id)
    except VersionedEntity.DoesNotExist:
        return False
    else:
        entity = version_entity.get_entity()
        resource_ids = entity.get(resource_name, '')
        if not resource_ids:
            return False
        if model_class.objects.filter(name=name, id__in=resource_ids.split(',')):
            return True
        return False


def validate_variable_inconfig(config):
    """校验配置文件中的变量名是否合法"""
    search_list = KEY_PATTERN.findall(json.dumps(config))
    search_keys = set(search_list)
    for ikey in search_keys:
        if not REAL_NUM_VAR_PATTERN.match(ikey):
            raise ValidationError(_('变量[{}]不合法, {}').format(ikey, NUM_VAR_ERROR_MSG))


def validate_res_config(config, resource_name, schema):
    err_prefix = '{resource_name} {suffix_msg}'.format(resource_name=resource_name, suffix_msg=_("配置信息格式错误"))
    try:
        json_validate(config, schema)
    except JsonValidationError as e:
        raise ValidationError(f'{err_prefix}:{e.message}')
    except SchemaError as e:
        raise ValidationError(f'{err_prefix}:{e}')


def validate_name_duplicate(data):
    resource_id = data.get('resource_id', None)
    version_id = data.get('version_id', None)
    if resource_id is None or version_id is None:
        return

    resource_name = data['resource_name']
    name = data['name']
    is_duplicate = is_name_duplicate(resource_name, resource_id, name, version_id)
    if is_duplicate:
        raise ValidationError(_('{}名称:{}已经在项目模板中被占用,请重新填写').format(resource_name, name))

File: bcs-services/cluster-resources/pkg/resource/form/validator/validator.go
/*
 * Tencent is pleased to support the open source community by making Blueking Container Service available.
 * Copyright (C) 2022 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * 	http://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under,
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Package validator xxx
package validator

import (
	"context"

	"github.com/TencentBlueKing/gopkg/collection/set"

	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/common/errcode"
	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/i18n"
	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/util/errorx"
	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/util/mapx"
	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/util/slice"
)

// FormDataValidator 表单数据校验器
// 目前只做基础的数据校验，表单校验主要还是前端按 Schema 规则完成，后续可支持后台校验 Schema 规则
type FormDataValidator struct {
	ctx        context.Context
	formData   map[string]interface{}
	apiVersion string
	kind       string
}

// New xxx
func New(ctx context.Context, formData map[string]interface{}, apiVersion, kind string) *FormDataValidator {
	return &FormDataValidator{ctx: ctx, formData: formData, apiVersion: apiVersion, kind: kind}
}

// Validate xxx
func (v *FormDataValidator) Validate() error {
	for _, f := range []func() error{
		// 1. 检查资源版本
		v.validateAPIVersion,
		// 2. 检查 Metadata 表单数据
		v.validateMetadata,
	} {
		if err := f(); err != nil {
			return err
		}
	}
	return nil
}

func (v *FormDataValidator) validateAPIVersion() error {
	supportedAPIVersions, ok := FormSupportedResAPIVersion[v.kind]
	if !ok {
		return errorx.New(errcode.Unsupported, i18n.GetMsg(v.ctx, "资源类型 `%s` 不支持表单化"), v.kind)
	}
	if !slice.StringInSlice(v.apiVersion, supportedAPIVersions) {
		return errorx.New(
			errcode.Unsupported,
			i18n.GetMsg(v.ctx, "资源类型 %s APIVersion %s 不在受支持的版本列表 %v 中，请改用 Yaml 模式而非表单化"),
			v.kind, v.apiVersion, supportedAPIVersions,
		)
	}
	return nil
}

func (v *FormDataValidator) validateMetadata() error {
	// 检查是否存在重复的 Label Key
	if err := checkDuplicateKey(
		mapx.GetList(v.formData, "metadata.labels"),
		i18n.GetMsg(v.ctx, "标签有重复的键，请检查"),
	); err != nil {
		return err
	}
	// 检查是否存在重复的 Annotations Key
	if err := checkDuplicateKey(
		mapx.GetList(v.formData, "metadata.annotations"),
		i18n.GetMsg(v.ctx, "注解有重复的键，请检查"),
	); err != nil {
		return err
	}
	return nil
}

// checkDuplicateKey 检查 k-v 列表中是否存在重复的 key
func checkDuplicateKey(kvList []interface{}, errMsg string) error {
	keys := set.NewStringSet()
	for _, kv := range kvList {
		k := kv.(map[string]interface{})["key"].(string)
		if keys.Has(k) {
			return errorx.New(errcode.ValidateErr, errMsg)
		}
		keys.Add(k)
	}
	return nil
}
